pkg/prometheus/promcfg_test.go

```diff
@@ -901,7 +911,7 @@ func TestK8SSDConfigGeneration(t *testing.T) {
 			sm.Namespace,
 			tc.apiServerConfig,
 			tc.store,
-			kubernetesSDRoleEndpoint,
+			tc.role,
```

Comment:
This check is valid, so we can keep it. However, it doesn't actually test the change introduced in this PR.

Looking at the implementation in this PR, the change happens at the `sync()` function of Prometheus/PrometheusAgent operator. Basically, `sync()` checks the value of `ServiceDiscoveryRole`, then set `endpointSliceSupported` accordingly. Meanwhile, this check is actually calling `generateK8SSDConfig()`, which is not a change in this PR.

So to check the change of this PR, could you add a new test to verify that `sync()` sets `endpointSliceSupported` according to the value of `ServiceDiscoveryRole`? Thankfully we already had [TestServiceMonitorWithEndpointSliceEnable](https://github.com/prometheus-operator/prometheus-operator/blob/ad006f4d607c99062d66936a8c02dd8bbaa6b40e/pkg/prometheus/promcfg_test.go#L1535C6-L1535C47) that verifies EndpointSlice is correctly used if `endpointSliceSupported == true`, so we don't need to care about this later step.

Reply:
Hello @haanhvu , thank you for pointing out that an existing function `TestServiceMonitorWithEndpointSliceEnable` exists which tests that a scrape configuration is properly generated if EndpointSlices are supported.

I am having trouble finding where in the code we have an example of the `sync()` function being tested. As the function is quite large it would probably require an integration test or extensive mocking. You are correct that the core logic of this PR is in the reconciliation loop evaluating if both 1) the Kubernetes API supports EndpointSlices and 2) the user has configured their prometheus-operator to use EndpointSlices for service discovery. **Would you be able to push me in the correct direction to test the reconciliation loop of Prometheus?**

The key lines I can identify are:

1. Where `sync()` is setting the `endpointSliceSupported` field on the `Operator` object: https://github.com/thaniri/prometheus-operator/blob/1da76da45e78597f973ba1341780c2e406786fd2/pkg/prometheus/server/operator.go#L813-L825
2. The function that creates new `Operator` objects: https://github.com/thaniri/prometheus-operator/blob/1da76da45e78597f973ba1341780c2e406786fd2/pkg/prometheus/server/operator.go#L123
3. Where the `New()` function is called from: https://github.com/thaniri/prometheus-operator/blob/1da76da45e78597f973ba1341780c2e406786fd2/cmd/operator/main.go#L352
4. Where the `Operator` object actually starts https://github.com/thaniri/prometheus-operator/blob/1da76da45e78597f973ba1341780c2e406786fd2/cmd/operator/main.go#L531-L533

I am envisioning instantiating an `Operator` and calling the `sync` function on it to see if the `endpointSliceSupported` field on the Operator actually gets updated.

**If there is no current testing on the `sync()` function, could that perhaps be its own issue?**

Reply:
Let me think how we can test this change as simply as we can. @simonpasquier @ArthurSens any ideas?

Reply:
@simonpasquier recommended moving out the core logic of this PR into `NewConfigGenerator()` in [this comment](https://github.com/prometheus-operator/prometheus-operator/pull/6672#pullrequestreview-2130701946). This is a very good suggestion as we will be able to easily unit test this.

I have modified TestServiceMonitorWithEndpointSliceEnable to account for this logic in [8ca42f5](https://github.com/prometheus-operator/prometheus-operator/pull/6672/commits/8ca42f506ccd1d3ae786aee76c004c9c65dee7d0)